src/engine/LocalVocab.h

```diff
@@ -24,12 +28,28 @@ class LocalVocab {
  private:
   using Entry = LocalVocabEntry;
   using LiteralOrIri = LocalVocabEntry;
+
+  // A functor that calculates the memory size of an IRI or Literal.
+  // This struct defines an operator() that takes a `LiteralOrIri` object and
+  // returns its dynamic memory usage in bytes.
+  struct IriSizeGetter {
+    ad_utility::MemorySize operator()(
+        const ad_utility::triple_component::LiteralOrIri& literalOrIri) {
+      return ad_utility::MemorySize::bytes(
+          literalOrIri.getDynamicMemoryUsage());
+    }
+  };
+
   // A map of the words in the local vocabulary to their local IDs. This is a
   // node hash map because we need the addresses of the words (which are of type
   // `LiteralOrIri`) to remain stable over their lifetime in the hash map
   // because we hand out pointers to them.
-  using Set = absl::node_hash_set<LiteralOrIri>;
-  std::shared_ptr<Set> primaryWordSet_ = std::make_shared<Set>();
+  using Set =
+      ad_utility::CustomHashSetWithMemoryLimit<LiteralOrIri, IriSizeGetter>;
+  ad_utility::detail::AllocationMemoryLeftThreadsafe limit_;
+  std::shared_ptr<Set> primaryWordSet_;
+
+  IriSizeGetter sizeGetter;
```

Comment:
I don't think you need a name for this variable, the size getter is default-constructible, so you don't need to specify it at all, as the default argument for the Set constructor works in this case.